provider/postgis/util.go

```diff
@@ -169,30 +155,63 @@ func replaceTokens(sql string, lyr *Layer, tile provider.Tile, withBuffer bool)
 	// replace query string tokens
 	z, x, y := tile.ZXY()
 	tokenReplacer := strings.NewReplacer(
-		bboxToken, bbox,
-		zoomToken, strconv.FormatUint(uint64(z), 10),
-		zToken, strconv.FormatUint(uint64(z), 10),
-		xToken, strconv.FormatUint(uint64(x), 10),
-		yToken, strconv.FormatUint(uint64(y), 10),
-		idFieldToken, lyr.IDFieldName(),
-		geomFieldToken, lyr.GeomFieldName(),
-		geomTypeToken, geoType,
-		scaleDenominatorToken, strconv.FormatFloat(scaleDenominator, 'f', -1, 64),
-		pixelWidthToken, strconv.FormatFloat(pixelWidth, 'f', -1, 64),
-		pixelHeightToken, strconv.FormatFloat(pixelHeight, 'f', -1, 64),
+		config.BboxToken, bbox,
+		config.ZoomToken, strconv.FormatUint(uint64(z), 10),
+		config.ZToken, strconv.FormatUint(uint64(z), 10),
+		config.XToken, strconv.FormatUint(uint64(x), 10),
+		config.YToken, strconv.FormatUint(uint64(y), 10),
+		config.ScaleDenominatorToken, strconv.FormatFloat(scaleDenominator, 'f', -1, 64),
+		config.PixelWidthToken, strconv.FormatFloat(pixelWidth, 'f', -1, 64),
+		config.PixelHeightToken, strconv.FormatFloat(pixelHeight, 'f', -1, 64),
+		config.IdFieldToken, lyr.IDFieldName(),
+		config.GeomFieldToken, lyr.GeomFieldName(),
+		config.GeomTypeToken, geoType,
 	)
 
 	uppercaseTokenSQL := uppercaseTokens(sql)
 
 	return tokenReplacer.Replace(uppercaseTokenSQL), nil
 }
 
-var tokenRe = regexp.MustCompile("![a-zA-Z0-9_-]+!")
+// replaceParams substitutes configured query parameter tokens for their values
+// within the provided SQL string
+func replaceParams(params map[string]provider.QueryParameter, sql string, args *[]interface{}) string {
+	if params == nil {
+		return sql
+	}
+
+	for _, token := range config.ParameterTokenRegexp.FindAllString(sql, -1) {
+		param := params[token]
+
+		// Replace every ? in the param's SQL with a positional argument
+		paramSQL := ""
+		for _, c := range param.SQL {
+			if c == '?' {
+				*args = append(*args, param.Value)
+				paramSQL = paramSQL + "$" + fmt.Sprint(len(*args))
```

Comment:
Or:
```suggestion
				paramSQL += paramSQL + fmt.Sprintf("$%d", len(*args)))
```

Reply:
If I'm reading your suggestion right, it will desugar into:
```go
paramSQL = paramSQL + paramSQL + fmt.Sprintf("$%d", len(*args)))
```

Guess the second `paramSQL` should not be there ;)

Reply:
Ah, yeah, it was supposed to be:
```go
paramSQL += fmt.Sprintf("$%d", len(*args)))
```
😅